File: data_structures/heap.cpp
#include <iostream>
#include <vector>
#include <stdexcept>
#include <algorithm>
#include <random>
#include <chrono>
#include <iomanip>
using namespace std;
using namespace std::chrono;

class Heap {
    private:
        vector<int> arr;
        int capacity;
        int parent(int i) {return (i - 1) / 2;}
        int leftChild(int i) {return 2 * i + 1;}
        int rightChild(int i) {return 2 * i + 2;}
        
        // Heapify up (for insertion)
        void heapifyUp(int i) {
            while (i > 0 && arr[parent(i)] < arr[i]) {
                swap(arr[i], arr[parent(i)]);
                i = parent(i);
            }
        }
        
        // Heapify down (for deletion)
        void heapifyDown(int i) {
            int maxIndex = i;
            int left = leftChild(i);
            int right = rightChild(i);
            
            if (left < arr.size() && arr[left] > arr[maxIndex]) {
                maxIndex = left;
            }
            
            if (right < arr.size() && arr[right] > arr[maxIndex]) {
                maxIndex = right;
            }
            
            if (i != maxIndex) {
                swap(arr[i], arr[maxIndex]);
                heapifyDown(maxIndex);
            }
        }
        
    public:
        // Constructor
        Heap(int cap = 100) : capacity(cap) {
            arr.reserve(capacity);
        }
        
        // Insert an element into the heap
        void insert(int value) {
            if (arr.size() >= capacity) {
                throw overflow_error("Heap is full");
            }
            
            arr.push_back(value);
            heapifyUp(arr.size() - 1);
        }
        
        // Extract the maximum element (root)
        int extractMax() {
            if (arr.empty()) {
                throw underflow_error("Heap is empty");
            }
            
            int maxValue = arr[0];
            arr[0] = arr.back();
            arr.pop_back();
            
            if (!arr.empty()) {
                heapifyDown(0);
            }
            
            return maxValue;
        }
        
        // Get the maximum element without removing it
        int getMax() const {
            if (arr.empty()) {
                throw underflow_error("Heap is empty");
            }
            return arr[0];
        }
        
        // Get the current size of the heap
        int size() const {
            return arr.size();
        }
        
        // Check if heap is empty
        bool isEmpty() const {return arr.empty();}
        
        // Build heap from an array (heapify)
        void buildHeap(const vector<int>& elements) {
            arr = elements;
            
            // Start from the last non-leaf node and heapify down
            for (int i = (arr.size() / 2) - 1; i >= 0; i--) {
                heapifyDown(i);
            }
        }
        
        // Display the heap
        void display() const {
            if (arr.empty()) {
                cout << "Heap is empty" << endl;
                return;
            }
            
            cout << "Heap elements: ";
            for (int val : arr) {
                cout << val << " ";
            }
            cout << endl;
        }
        
        // Heap sort (returns sorted array in descending order)
        vector<int> heapSort() {
            vector<int> sorted;
            Heap tempHeap = *this;
            
            while (!tempHeap.isEmpty()) {
                sorted.push_back(tempHeap.extractMax());
            }
            
            return sorted;
        }
};

// Generate random data
vector<int> generateRandomData(int n, int seed = 42) {
    vector<int> data;
    mt19937 gen(seed);
    uniform_int_distribution<> dis(1, 100000);
    
    for (int i = 0; i < n; i++) {
        data.push_back(dis(gen));
    }
    
    return data;
}

// Benchmark insertion
void benchmarkInsertion(int n) {
    vector<int> data = generateRandomData(n);
    
    // Heap insertion
    auto start = high_resolution_clock::now();
    Heap heap;
    for (int val : data) {
        heap.insert(val);
    }
    auto end = high_resolution_clock::now();
    auto heapTime = duration_cast<microseconds>(end - start).count();
    
    // List insertion (unsorted)
    start = high_resolution_clock::now();
    vector<int> list;
    for (int val : data) {
        list.push_back(val);
    }
    end = high_resolution_clock::now();
    auto listTime = duration_cast<microseconds>(end - start).count();
    
    // Sorted list insertion (maintaining sorted order)
    start = high_resolution_clock::now();
    vector<int> sortedList;
    for (int val : data) {
        auto pos = lower_bound(sortedList.begin(), sortedList.end(), val, greater<int>());
        sortedList.insert(pos, val);
    }
    end = high_resolution_clock::now();
    auto sortedListTime = duration_cast<microseconds>(end - start).count();
    
    cout << "Insertion (" << n << " elements):" << endl;
    cout << "  Heap:             " << setw(10) << heapTime << " μs" << endl;
    cout << "  List (unsorted):  " << setw(10) << listTime << " μs" << endl;
    cout << "  List (sorted):    " << setw(10) << sortedListTime << " μs" << endl;
    cout << "  Heap speedup vs sorted list: " << fixed << setprecision(2) 
         << (double)sortedListTime / heapTime << "x" << endl << endl;
}

// Benchmark finding maximum
void benchmarkFindMax(int n) {
    vector<int> data = generateRandomData(n);
    
    // Build heap
    Heap heap;
    for (int val : data) {
        heap.insert(val);
    }
    
    // Build list
    vector<int> list = data;
    
    // Heap getMax
    auto start = high_resolution_clock::now();
    for (int i = 0; i < 1000; i++) {
        int max = heap.getMax();
    }
    auto end = high_resolution_clock::now();
    auto heapTime = duration_cast<nanoseconds>(end - start).count() / 1000.0;
    
    // List max_element
    start = high_resolution_clock::now();
    for (int i = 0; i < 1000; i++) {
        auto maxIt = max_element(list.begin(), list.end());
    }
    end = high_resolution_clock::now();
    auto listTime = duration_cast<nanoseconds>(end - start).count() / 1000.0;
    
    cout << "Find Maximum (" << n << " elements, averaged over 1000 calls):" << endl;
    cout << "  Heap:  " << setw(10) << fixed << setprecision(2) << heapTime << " ns" << endl;
    cout << "  List:  " << setw(10) << fixed << setprecision(2) << listTime << " ns" << endl;
    cout << "  Heap speedup: " << fixed << setprecision(2) 
         << listTime / heapTime << "x" << endl << endl;
}

// Benchmark extraction of maximum
void benchmarkExtractMax(int n) {
    vector<int> data = generateRandomData(n);
    
    // Build heap
    Heap heap;
    for (int val : data) {
        heap.insert(val);
    }
    
    // Build list
    vector<int> list = data;
    
    // Heap extractMax
    auto start = high_resolution_clock::now();
    while (!heap.isEmpty()) {
        heap.extractMax();
    }
    auto end = high_resolution_clock::now();
    auto heapTime = duration_cast<microseconds>(end - start).count();
    
    // List extract max (find and erase)
    start = high_resolution_clock::now();
    while (!list.empty()) {
        auto maxIt = max_element(list.begin(), list.end());
        list.erase(maxIt);
    }
    end = high_resolution_clock::now();
    auto listTime = duration_cast<microseconds>(end - start).count();
    
    cout << "Extract All Maximum Elements (" << n << " elements):" << endl;
    cout << "  Heap:  " << setw(10) << heapTime << " μs" << endl;
    cout << "  List:  " << setw(10) << listTime << " μs" << endl;
    cout << "  Heap speedup: " << fixed << setprecision(2) 
         << (double)listTime / heapTime << "x" << endl << endl;
}

// Benchmark building from existing data
void benchmarkBuildHeap(int n) {
    vector<int> data = generateRandomData(n);
    
    // Build heap from array
    auto start = high_resolution_clock::now();
    Heap heap;
    heap.buildHeap(data);
    auto end = high_resolution_clock::now();
    auto heapTime = duration_cast<microseconds>(end - start).count();
    
    // Sort list
    vector<int> list = data;
    start = high_resolution_clock::now();
    sort(list.begin(), list.end(), greater<int>());
    end = high_resolution_clock::now();
    auto listTime = duration_cast<microseconds>(end - start).count();
    
    cout << "Build/Sort from " << n << " elements:" << endl;
    cout << "  Heap (buildHeap):  " << setw(10) << heapTime << " μs" << endl;
    cout << "  List (sort):       " << setw(10) << listTime << " μs" << endl;
    cout << "  Heap speedup: " << fixed << setprecision(2) 
         << (double)listTime / heapTime << "x" << endl << endl;
}

// Benchmark priority queue operations
void benchmarkPriorityQueue(int n) {
    vector<int> data = generateRandomData(n);
    
    // Heap: insert all, then extract top k elements
    int k = min(100, n);
    auto start = high_resolution_clock::now();
    Heap heap;
    for (int val : data) {
        heap.insert(val);
    }
    vector<int> topK;
    for (int i = 0; i < k; i++) {
        topK.push_back(heap.extractMax());
    }
    auto end = high_resolution_clock::now();
    auto heapTime = duration_cast<microseconds>(end - start).count();
    
    // List: sort and get top k
    start = high_resolution_clock::now();
    vector<int> list = data;
    sort(list.begin(), list.end(), greater<int>());
    vector<int> topKList(list.begin(), list.begin() + k);
    end = high_resolution_clock::now();
    auto listTime = duration_cast<microseconds>(end - start).count();
    
    cout << "Priority Queue: Get top " << k << " from " << n << " elements:" << endl;
    cout << "  Heap:  " << setw(10) << heapTime << " μs" << endl;
    cout << "  List:  " << setw(10) << listTime << " μs" << endl;
    cout << "  Heap speedup: " << fixed << setprecision(2) 
         << (double)listTime / heapTime << "x" << endl << endl;
}

int main() {
    cout << "=====================================" << endl;
    cout << "    HEAP vs LIST BENCHMARK" << endl;
    cout << "=====================================" << endl << endl;
    
    vector<int> sizes = {100, 1000, 5000, 10000};
    
    for (int n : sizes) {
        cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << endl;
        cout << "  Dataset Size: " << n << " elements" << endl;
        cout << "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━" << endl << endl;
        
        benchmarkInsertion(n);
        benchmarkFindMax(n);
        benchmarkExtractMax(n);
        benchmarkBuildHeap(n);
        benchmarkPriorityQueue(n);
    }
    
    cout << "=====================================" << endl;
    cout << "       BENCHMARK COMPLETED" << endl;
    cout << "=====================================" << endl << endl;
    
    cout << "Key Observations:" << endl;
    cout << "1. Heap insertion: O(log n) - Good for dynamic data" << endl;
    cout << "2. Heap getMax: O(1) - Instant access to maximum" << endl;
    cout << "3. Heap extractMax: O(log n) - Efficient priority queue" << endl;
    cout << "4. Heap buildHeap: O(n) - Faster than sorting" << endl;
    cout << "5. List is better for simple append operations" << endl;
    cout << "6. Heap is significantly better for priority queue operations" << endl;
    
    return 0;
}

File: data_structures/heap.py
class Heap:
    """
    Min Heap Implementation

    Variables:
        - heap: just a list
        - size: the size of the heap
        - index: the index of the node in the heap list

    Public Functions:
        - insert: inserts a key into the heap
        - extract_min: extracts the minimum element from the heap
        - build_heap: builds a heap from an array
        - display: displays the heap
        - heap_sort: returns a sorted array
    """
    def __init__(self):
        self.heap = []
        self.size = 0

    def _display_helper(self, index, indent):
        if index < self.size:
            print(' ' * indent + '-->', self.heap[index])

    def _parent(self, index): 
        return (index - 1) // 2
    
    def _left(self, index): 
        return 2 * index + 1
    
    def _right(self, index): 
        return 2 * index + 2
    
    def _swap(self, i, j): 
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def _heapify_up(self, index):
        if index == 0: return
        curr_val, parent_val = self.heap[index], self.heap[self._parent(index)]
        if curr_val < parent_val:
            self._swap(index, self._parent(index))
            self._heapify_up(self._parent(index))
        else: return

    def _heapify_down(self, index):
        if index >= self.size: return
        smallest = index
        left_idx = self._left(index)
        right_idx = self._right(index)
        
        if left_idx < self.size and self.heap[left_idx] < self.heap[smallest]:
            smallest = left_idx
        
        if right_idx < self.size and self.heap[right_idx] < self.heap[smallest]:
            smallest = right_idx
        
        if smallest != index:
            self._swap(index, smallest)
            self._heapify_down(smallest)

    def insert(self, key):
        self.heap.append(key)
        self.size += 1
        self._heapify_up(self.size - 1)

    def extract_min(self):
        if self.size == 0: return None
        if self.size == 1: return self.heap.pop()
        min_val = self.heap[0]
        self.heap[0] = self.heap.pop()
        self.size -= 1
        self._heapify_down(0)
        return min_val

    def build_heap(self, arr):
        self.heap = arr
        self.size = len(arr)
        for i in range(self.size // 2 - 1, -1, -1):
            self._heapify_down(i)
    
    def display(self, indent=0):
        if self.size == 0: return
        if indent == 0: print(self.heap[0])
        else: print(' ' * indent + '-->', self.heap[0])
        for i in range(1, self.size // 2):
            self._display_helper(self._left(i), indent + 1)
            self._display_helper(self._right(i), indent + 1)

    def heap_sort(self):
        sorted = []
        while self.size > 0:
            sorted.append(self.extract_min())
        return sorted
        

import random
import time

def generate_random_data(n):
    return [random.randint(1, 100000) for _ in range(n)]

def benchmark_insertion(n):
    data = generate_random_data(n)
    
    # Heap insertion
    heap = Heap()
    start = time.time()
    for val in data:
        heap.insert(val)
    end = time.time()
    heap_time = (end - start) * 1000
    
    # List insertion (unsorted)
    lst = []
    start = time.time()
    for val in data:
        lst.append(val)
    end = time.time()
    list_time = (end - start) * 1000
    
    # Sorted list insertion (maintaining sorted order)
    sorted_list = []
    start = time.time()
    for val in data:
        sorted_list.append(val)
        sorted_list.sort()
    end = time.time()
    sorted_list_time = (end - start) * 1000
    
    print(f"Insertion ({n} elements):")
    print(f"  Heap:            {heap_time:10.2f} ms")
    print(f"  List (unsorted): {list_time:10.2f} ms")
    print(f"  List (sorted):   {sorted_list_time:10.2f} ms")
    print(f"  Heap speedup vs sorted list: {sorted_list_time / heap_time:.2f}x\n")

def benchmark_find_min(n):
    data = generate_random_data(n)
    
    # Build heap
    heap = Heap()
    for val in data:
        heap.insert(val)
    
    # Build list
    lst = data.copy()
    
    # Heap get min
    iterations = 1000
    start = time.time()
    for _ in range(iterations):
        min_val = heap.heap[0] if heap.size > 0 else None
    end = time.time()
    heap_time = ((end - start) * 1000000) / iterations
    
    # List min
    start = time.time()
    for _ in range(iterations):
        min_val = min(lst) if lst else None
    end = time.time()
    list_time = ((end - start) * 1000000) / iterations
    
    print(f"Find Minimum ({n} elements, averaged over {iterations} calls):")
    print(f"  Heap: {heap_time:10.2f} ns")
    print(f"  List: {list_time:10.2f} ns")
    print(f"  Heap speedup: {list_time / heap_time:.2f}x\n")

def benchmark_extract_min(n):
    data = generate_random_data(n)
    
    # Build heap
    heap = Heap()
    for val in data:
        heap.insert(val)
    
    # Build list
    lst = data.copy()
    
    # Heap extract_min
    start = time.time()
    while heap.size > 0:
        heap.extract_min()
    end = time.time()
    heap_time = (end - start) * 1000
    
    # List extract min
    start = time.time()
    while lst:
        min_val = min(lst)
        lst.remove(min_val)
    end = time.time()
    list_time = (end - start) * 1000
    
    print(f"Extract All Minimum Elements ({n} elements):")
    print(f"  Heap: {heap_time:10.2f} ms")
    print(f"  List: {list_time:10.2f} ms")
    print(f"  Heap speedup: {list_time / heap_time:.2f}x\n")

def benchmark_build_heap(n):
    data = generate_random_data(n)
    
    # Build heap from array
    start = time.time()
    heap = Heap()
    heap.build_heap(data.copy())
    end = time.time()
    heap_time = (end - start) * 1000
    
    # Sort list
    start = time.time()
    lst = data.copy()
    lst.sort()
    end = time.time()
    list_time = (end - start) * 1000
    
    print(f"Build/Sort from {n} elements:")
    print(f"  Heap (build_heap): {heap_time:10.2f} ms")
    print(f"  List (sort):       {list_time:10.2f} ms")
    print(f"  Heap speedup: {list_time / heap_time:.2f}x\n")

def benchmark_priority_queue(n):
    data = generate_random_data(n)
    
    # Heap: insert all, then extract top k elements
    k = min(100, n)
    start = time.time()
    heap = Heap()
    for val in data:
        heap.insert(val)
    top_k = []
    for _ in range(k):
        top_k.append(heap.extract_min())
    end = time.time()
    heap_time = (end - start) * 1000
    
    # List: sort and get top k
    start = time.time()
    lst = data.copy()
    lst.sort()
    top_k_list = lst[:k]
    end = time.time()
    list_time = (end - start) * 1000
    
    print(f"Priority Queue: Get top {k} from {n} elements:")
    print(f"  Heap: {heap_time:10.2f} ms")
    print(f"  List: {list_time:10.2f} ms")
    print(f"  Heap speedup: {list_time / heap_time:.2f}x\n")

if __name__ == '__main__':
    print("=" * 40)
    print("    HEAP vs LIST BENCHMARK")
    print("=" * 40)
    print()
    
    sizes = [100, 1000, 5000, 10000]
    
    for n in sizes:
        print("=" * 40)
        print(f"  Dataset Size: {n} elements")
        print("=" * 40)
        print()
        
        benchmark_insertion(n)
        benchmark_find_min(n)
        benchmark_extract_min(n)
        benchmark_build_heap(n)
        benchmark_priority_queue(n)
    
    print("=" * 40)
    print("       BENCHMARK COMPLETED")
    print("=" * 40)
    print()
    
    print("Key Observations:")
    print("1. Heap insertion: O(log n) - Good for dynamic data")
    print("2. Heap get_min: O(1) - Instant access to minimum")
    print("3. Heap extract_min: O(log n) - Efficient priority queue")
    print("4. Heap build_heap: O(n) - Faster than sorting")
    print("5. List is better for simple append operations")
    print("6. Heap is significantly better for priority queue operations")